base-lite/ub/ub.go

```diff
@@ -297,6 +316,70 @@ func checkKafkaReady(minNumBroker string, timeout string, bootstrapServers strin
 	return invokeJavaCommand("io.confluent.admin.utils.cli.KafkaReadyCommand", jvmOpts, opts)
 }
 
+func waitForServer(host string, port int, timeout time.Duration) bool {
+	address := fmt.Sprintf("%s:%d", host, port)
+	startTime := time.Now()
+	connectTimeout := 5 * time.Second
+
+	for {
+		conn, err := net.DialTimeout("tcp", address, connectTimeout)
+		if err == nil {
+			_ = conn.Close()
+			return true
+		}
+		if time.Since(startTime) >= timeout {
+			return false
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func waitForHttp(urlString string, timeout time.Duration) error {
+	parsedUrl, err := url.Parse(urlString)
+	if err != nil {
+		err = fmt.Errorf("error in parsing url %q: %w", urlString, err)
+		return err
+	}
+
+	host := parsedUrl.Hostname()
+	portStr := parsedUrl.Port()
+
+	if len(host) == 0 {
+		host = "localhost"
+	}
+
+	if len(portStr) == 0 {
+		portMap := map[string]string{"http": "80", "https": "443"}
+		defaultPort, found := portMap[parsedUrl.Scheme]
+		if !found {
+			err = fmt.Errorf("no port specified and cannot infer port based on protocol (only http(s) supported)")
+			return err
+		}
+		portStr = defaultPort
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		err = fmt.Errorf("error in parsing port %q: %w", portStr, err)
+		return err
+	}
+
+	if !waitForServer(host, port, timeout) {
```

Comment:
This is serving 2 purposes
1. Extra check if server is up or not.
2. It will be helpful in more verbose error message, i.e, whether the failure was due to server not coming up, or server coming up but the url being unreachable.